### PR TITLE
feat: Don't Label Users as Quota Limited for N Days After They Exceed Their Quota

### PR DESCRIPTION
**Pull request by xrdt**
## Problem

N = 7. We had discussed making N = 3 because 7 might just be too many days. 

We would like to keep users' data for N days after they start to exceed their usage quotas. This will allow us to recover in a case where billing limits inadvertently become lower than current usage ([INC-144](https://posthog.slack.com/archives/C06E3RC442V)). This PR introduces a new redis cache and additional metadata on the cached usage dict in posthog to record this temporary state. 

Working on tests 🔨 

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Comment by xrdt** [comment]
We're going to keep this strategy in place as a failsafe to complement the manual, feature-flag based retention trigger: https://github.com/PostHog/posthog/pull/19996

File: ee/billing/test/test_quota_limiting.py [hunk 1]
```diff
@@ -76,25 +85,110 @@ def test_billing_rate_limit(self) -> None:
...
 
-        assert self.redis_client.zrange(f"{QUOTA_LIMITER_CACHE_KEY}events", 0, -1) == [
-            self.team.api_token.encode("UTF-8")
-        ]
-        assert self.redis_client.zrange(f"{QUOTA_LIMITER_CACHE_KEY}recordings", 0, -1) == []
-        assert self.redis_client.zrange(f"{QUOTA_LIMITER_CACHE_KEY}rows_synced", 0, -1) == []
+        with freeze_time("2021-01-25T22:09:14.252Z"):
+            # Should be data_retained until Feb 1 2021
+            quota_limited_orgs, data_retained_orgs = update_all_org_billing_quotas()
+            assert data_retained_orgs["events"] == {org_id: 1612137600}
```
Comment (raquelmsmith) [inline ee/billing/test/test_quota_limiting.py:93]: does this org_id change ever?

**Review by raquelmsmith — COMMENTED** [review]
Logic looks good here, just a couple questions about code hygiene. 

I wonder what is the best way to use this and keep an eye on it. Like should we have an alert of some sort when the number of teams who _would have_ been limited goes up?

File: ee/billing/test/test_quota_limiting.py (hunk 1 above)
Comment (xrdt) [reply ee/billing/test/test_quota_limiting.py:93]: This isn't the org_id but the time at which data retention for this org should end.

Comment (xrdt) [reply ee/billing/test/test_quota_limiting.py:93]: It's inconsistent with the other object for quota_limited_orgs...Will think about it.

File: ee/billing/quota_limiting.py [hunk 2]
```diff
@@ -78,19 +91,40 @@ def org_quota_limited_until(organization: Organization, resource: QuotaResource)
...
+    # Either wasn't set or was set in the previous biling period
+    if (
+        not data_retained_until
+        or round((datetime.fromtimestamp(data_retained_until) - timedelta(days=7)).timestamp()) < billing_period_start
+    ):
+        data_retained_until = round((today + timedelta(days=7)).timestamp())
+        summary["retained_period_end"] = data_retained_until
+        needs_save = True
+
+    if billing_period_end:
```
[the remark this answers is not in the document]
Comment (xrdt) [reply ee/billing/quota_limiting.py:126]: I can't imagine why we wouldn't have one. This is here simply to be consistent with the previous check `if is_quota_limited and billing_period_end:`

File: ee/billing/test/test_quota_limiting.py (hunk 1 above)
Comment (xrdt) [reply ee/billing/test/test_quota_limiting.py:93]: The org_id in this key gets replaced with the organization.id used in this test so the value we're asserting against looks more like {'uuid123123123': 1612137600}

**Comment by xrdt** [comment]
It seems like we'd be most interested in keeping tabs on whether certain orgs are incorrectly having data retained. That would imply we're not correctly inserting and removing them from the data-retention redis set. I think we should create a dashboard similar to the existing quota limiting one, or simply add graphs for data retention to that dashboard.

File: ee/billing/test/test_quota_limiting.py (hunk 1 above)
Comment (raquelmsmith) [reply ee/billing/test/test_quota_limiting.py:93]: ohh I get it 👍

**Review by raquelmsmith — APPROVED** [review]

**Pull request merged**